File: src/D3D11/D3D11Buffer.cpp
#include "D3D11Buffer.h"

#include "D3D11Utils.h"

namespace grabs::D3D11
{
    D3D11Buffer::D3D11Buffer(ID3D11Device* device, const BufferDescription& description, void* data)
    {
        UINT bindFlags = 0;
        switch (description.Type)
        {
            case BufferType::Vertex:
                bindFlags = D3D11_BIND_VERTEX_BUFFER;
                break;
            case BufferType::Index:
                bindFlags = D3D11_BIND_INDEX_BUFFER;
                break;
            case BufferType::Constant:
                bindFlags = D3D11_BIND_CONSTANT_BUFFER;
                break;
        }

        const D3D11_BUFFER_DESC desc
        {
            .ByteWidth = description.Size,
            .Usage = description.Dynamic ? D3D11_USAGE_DYNAMIC : D3D11_USAGE_DEFAULT,
            .BindFlags = bindFlags,
            .CPUAccessFlags = 0
        };

        const D3D11_SUBRESOURCE_DATA subData { .pSysMem = data };
        D3D11_CHECK_RESULT(device->CreateBuffer(&desc, data ? &subData : nullptr, &Buffer));
    }

    D3D11Buffer::~D3D11Buffer()
    {
        Buffer->Release();
    }
}

File: src/grabs.Graphics/D3D11/D3D11Buffer.cs
using System.Diagnostics.CodeAnalysis;
using TerraFX.Interop.DirectX;
using static grabs.Graphics.D3D11.D3D11Result;
using static TerraFX.Interop.DirectX.D3D11_BIND_FLAG;
using static TerraFX.Interop.DirectX.D3D11_USAGE;

namespace grabs.Graphics.D3D11;

[SuppressMessage("Interoperability", "CA1416:Validate platform compatibility")]
internal sealed unsafe class D3D11Buffer : Buffer
{
    public readonly ID3D11Buffer* Buffer;

    public D3D11Buffer(ID3D11Device* device, in BufferDescription description, void* data)
    {
        D3D11_BIND_FLAG bindFlags = description.Type switch
        {
            BufferType.Vertex => D3D11_BIND_VERTEX_BUFFER,
            BufferType.Index => D3D11_BIND_INDEX_BUFFER,
            BufferType.Constant => D3D11_BIND_CONSTANT_BUFFER,
            _ => throw new ArgumentOutOfRangeException()
        };

        D3D11_BUFFER_DESC desc = new()
        {
            BindFlags = (uint) bindFlags,
            ByteWidth = description.Size,
            Usage = description.Dynamic ? D3D11_USAGE_DYNAMIC : D3D11_USAGE_DEFAULT
        };

        D3D11_SUBRESOURCE_DATA resData = new()
        {
            pSysMem = data
        };

        fixed (ID3D11Buffer** buffer = &Buffer)
            CheckResult(device->CreateBuffer(&desc, data == null ? null : &resData, buffer), "Create buffer");
    }
    
    public override void Dispose()
    {
        Buffer->Release();
    }
}